lib/portlayer/storage/nfs/volume_test.go

```diff
@@ -0,0 +1,516 @@
+	return MockTarget{dirPath: path}
+}
+
+func (v MockTarget) Open(path string) (io.ReadCloser, error) {
+	return os.Open(path)
+}
+
+func (v MockTarget) OpenFile(path string, mode os.FileMode) (io.ReadWriteCloser, error) {
+	return os.OpenFile(path, os.O_RDWR, mode)
+}
+
+func (v MockTarget) Create(path string, perm os.FileMode) (io.ReadWriteCloser, error) {
+	return os.Create(path)
+}
+
+func (v MockTarget) MkDir(path string, perm os.FileMode) ([]byte, error) {
+	return nil, os.Mkdir(path, perm)
+}
+
+func (v MockTarget) RemoveAll(path string) error {
+	return os.RemoveAll(path)
+}
+
+func (v MockTarget) ReadDir(path string) ([]os.FileInfo, error) {
+	dir, err := os.Open(path)
+	if err != nil {
+		return nil, err
+	}
+
+	return dir.Readdir(0)
+}
+
+func (v MockTarget) Lookup(path string) (os.FileInfo, error) {
+	return os.Stat(path)
+}
+
+// MOCK MOUNT STRUCT AND IMPL
+
+type MockMount struct {
+}
+
+func (m MockMount) Mount(target *url.URL) (NFSTarget, error) {
+	return NewMocktarget(target.Path), nil
+}
+
+func (m MockMount) Unmount(target NFSTarget) error {
+	return nil
+}
+
+func TestMain(m *testing.M) {
+	testPath := path.Join(os.TempDir(), nfsTestDir)
+	os.Mkdir(testPath, 0755)
+	result := m.Run()
+	os.RemoveAll(testPath)
+	os.Exit(result)
+}
+
+func TestSimpleVolumeStoreOperations(t *testing.T) {
+	mockMount := MockMount{}
+	testVolName := "testVolume"
+	dirpath := path.Join(os.TempDir(), nfsTestDir)
+
+	targetURL := url.URL{Path: dirpath}
+	op := trace.NewOperation(context.TODO(), "TestOp")
+
+	//Create a Volume Store
+	vs, err := NewVolumeStore(op, "testStore", &targetURL, mockMount)
+	if !assert.Nil(t, err, "Failed during call to NewVolumeStore with err (%s)", err) {
+		return
+	}
+
+	_, err = os.Stat(dirpath)
+	if !assert.Nil(t, err, "Could not find the initial volume store directory after creation of volume store. err (%s)", err) {
+		return
+	}
+
+	if !assert.NotNil(t, vs, "Volume Store created with nil err, but return is also nil") {
+		return
+	}
+
+	info := make(map[string][]byte)
+	testInfoKey := "junk"
+	info[testInfoKey] = make([]byte, 20)
+
+	file, err := os.Open(dirpath)
+	contents, err := file.Readdir(0)
+	t.Logf("%s\n", contents)
+
+	//Create a Volume
+	vol, err := vs.VolumeCreate(op, testVolName, vs.Target, 0 /*we do not use this*/, info)
+
+	if !assert.Nil(t, err, "Failed during call to VolumeCreate with err (%s)", err) {
+		return
+	}
+
+	if !assert.Equal(t, testVolName, vol.ID, "expected volume ID (%s) got ID (%s)", testVolName, vol.ID) {
+		return
+	}
+
+	_, ok := vol.Info[testInfoKey]
+	if !assert.True(t, ok, "TestInfoKey did not exist in the return metadata map") {
+		return
+	}
+
+	//Check Metadata Pathing
+	metadataPath := path.Join(dirpath, metadataDir)
+	volumePath := path.Join(dirpath, VolumesDir)
+
+	metaFilesDir, err := os.Open(metadataPath)
+	if !assert.Nil(t, err, "opening the metadata directory failed with err (%s)", err) {
+		return
+	}
+
+	volumeDir, err := os.Open(volumePath)
```

Comment:
I went back and did this in a lot of places. Previously I was just closing everything I opened at the bottom of the function call.